tools/topology/sof-ehl-rt5660.m4

```diff
@@ -0,0 +1,193 @@
+#
+# Topology for ELKHARTLAKE with rt5660 codec
+#
+
+# Include topology builder
+include(`utils.m4')
+include(`dai.m4')
+include(`pipeline.m4')
+include(`ssp.m4')
+
+# Include TLV library
+include(`common/tlv.m4')
+
+# Include Token library
+include(`sof/tokens.m4')
+
+# Include Elkhartlake DSP configuration
+include(`platform/intel/ehl.m4')
+include(`platform/intel/dmic.m4')
+
+DEBUG_START
+
+#
+# Define the pipelines
+#
+# PCM0 ----> volume -----> SSP0 BE dailink 0
+# PCM1 <---- volume <----- DMIC48k (dmic48k, BE dailink 1)
+
+ifelse(HDMI, `1',
+`
+# PCM2 ----> volume -----> iDisp1 (HDMI/DP playback, BE link 5)
+# PCM3 ----> volume -----> iDisp2 (HDMI/DP playback, BE link 6)
+# PCM4 ----> volume -----> iDisp3 (HDMI/DP playback, BE link 7)
+# PCM5 ----> volume -----> iDisp3 (HDMI/DP playback, BE link 8)
+')
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     frames, deadline, priority, core)
+
+# Low Latency playback pipeline 1 on PCM 0 using max 2 channels of s32le.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
+	1, 0, 2, s32le,
+	1000, 0, 0,
+	48000, 48000, 48000)
```

Comment:
This looks wrong, the last three args for PCM_ADD are not rate.

Reply:
Missing time_domain and sched_comp if I follow the arguments. It is indeed wrong to not define them but I believe somehow it manages to do something (sane defaults for at least my platform) even with those set as `' (I had them as such before and patched them up later).

Reply:
@lgirdwood @paulstelian97 
The definition should be right. As Paulstelian mentioned my patch doesn't define the last two arguments as the last tow arguments are not used.
The last "PIPELINE_PCM_ADD" update is made in 77b6db8 by @juimonen. My patch follows that patch's sample writing.

Reply:
@libinyang Even if that particular pipeline doesn't use them you should still put some sane values in case someone comes and changes only the pipeline file but not the parameters. So even if it works you should be more careful and make the code more reliable when it will get changed.

Reply:
@lgirdwood @paulstelian97 Do you have some recommendation for the default value? I checked the other m4 files and didn't find an example except some m4 files really use the parameters. I'm not sure set that value to this patch is OK or not. And as there are lots of m4 doesn't set these values, I do think there should be a separate patch to do that. Let's not fix the common issue in this patch. What do you think?

Reply:
@libinyang this wont merge unless the correct values are used. @juimonen is fixing the values for the existing topology files. Please align with @juimonen for correct values

Reply:
@lgirdwood I think the pipeline values look good, I only added missing TIMER/DMA values to DAI_ADD.

Reply:
@juimonen I think @lgirdwood said that this PR should also have those parameters set.

Reply:
@juimonen Thanks for the confirm.

@paulstelian97 @lgirdwood OK. I will add that for the missing value of DAI_ADD.

Reply:
@libinyang I think you didn't update here?

Reply:
@lgirdwood @paulstelian97 Do we need update here? Jaska said the pipeline looks good.

Reply:
@libinyang Seems like TIME_DOMAIN and SCHED_COMP are unset. I'd argue that TIME_DOMAIN may still need to be set, probably to the same value as the pipeline itself (either DMA or TIMER)

Reply:
@paulstelian97 TIME_DOMAIN is only "needed" in DAI_ADD. _but_  for PIPELINE_ADD you need to define SCHED_COMP and TIME_DOMAIN if you have a "branched" pipeline, so like 2 pipelines connected to 1 dai, because otherwise the 2nd pipeline is not tied to any scheduling component. When you have 1 pipeline to 1 dai connections the scheduler is implicit in the pipeline (as dai defines it).

This is why you see in these "branched" topologies that the 2nd pipeline is defined _after_ the dai, because the dai defines the scheduler, which you then use in 2nd pipeline. 

Simple isnt' it? :) Maybe we could mandate it in every pipeline, so it would be more clear...

So in this case I see on only 1 to 1 mappings, so these should not be needed in pipeline definitions.

Reply:
so I guess in this case you _could_ define it, but as the dai definitions come after, the define would be empty and I'm not sure would it break something, or just work.

Always if you see this SCHED_COMP in pipeline it means that it is something special, so it is branched (has probably mixer component), or like the wov pipelines where the scheduler is not obvious, or some other more "complex" pipeline where it needs to be explicitly defined.

This was added at some point to help the firmware to understand how to process the connected pipelines. So for example in the mixer pipeline it should be pretty obvious who is the scheduler, but we might have some cases where it needs to be manually defined. And not to blame firmware only, branched/cyclic or in any way more complex pipelines are an issue also on the kernel/asoc side.

Reply:
@juimonen Alright I suppose. @lgirdwood Is this still an issue?

Reply:
Ok, looks good - @juimonen can you document these assumptions on sof-docs.